### PR TITLE
Set of GIF manipulation functions

### PR DESCRIPTION
**Pull request by cachemoi**
Functions to manipulate GIFs and autogenerate assemblies to make them with bacteria

### CODE REVIEW

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go [hunk 1]
```diff
@@ -263,6 +266,26 @@ func OpenFile(file wtype.File) (nrgba *goimage.NRGBA, err error) {
 	return nrgba, nil
 }
 
+func OpenGIF(file wtype.File)(GIF *gif.GIF, err error){
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go]: godoc: `OpenGif ...`

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go [hunk 2]
```diff
@@ -1351,6 +1401,20 @@ func (p AnthaPalette) Index(c color.Color) int {
 	return ret
 }
 
+// Returns a bool indicating if the AnthaColor is already in the AnthaPalette
+func (p AnthaPalette) CheckPresence(c AnthaColor) (ok bool) {
+
+	ok = false
+
+	for _, pc := range p.AnthaColors{
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go]: in general, gofmt all your files.

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go [hunk 3]
```diff
@@ -1351,6 +1401,20 @@ func (p AnthaPalette) Index(c color.Color) int {
...
+		if reflect.DeepEqual(pc,c) {
+			ok =  true
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go]: return true

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go [hunk 4]
```diff
@@ -1351,6 +1401,20 @@ func (p AnthaPalette) Index(c color.Color) int {
...
+		}
+	}
+
+	return ok
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go]: return false

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go [hunk 5]
```diff
@@ -1223,6 +1271,7 @@ type AnthaImg struct {
 
 //Set of types to use antha with biological colors
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go]: godoc style: // A LivingColor ...

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go [hunk 6]
```diff
@@ -1402,28 +1466,101 @@ func sqDiff(x, y uint32) uint32 {
 	return (d * d) >> 2
 }
 
-//---------------------------------------------------
-//Object constructors
-//---------------------------------------------------
+// Returns a bool indicating if the LivingColor is already in the LivingPalette
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go]: godoc style: `// CheckPresence ...`

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go [hunk 7]
```diff
@@ -1402,28 +1466,101 @@ func sqDiff(x, y uint32) uint32 {
...
-		panic(err)
+	return ok
+}
+
+// Returns a bool indicating if the LivingPixel is the same as the given one
+func (p1 LivingPix) Compare(p2 LivingPix) (same bool) {
+
+	if reflect.DeepEqual(p1,p2) {
+		return true
+	}else{
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go]: gofmt

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go [hunk 8]
```diff
@@ -1402,28 +1466,101 @@ func sqDiff(x, y uint32) uint32 {
...
 
-//Object constructor for a LivingColor with default settings
-func MakeLivingColor(ctx context.Context, color *color.NRGBA, seq string) (livingColor *LivingColor) {
+	ok = false
 
-	//generating DNA sequence object
-	DNASequence := wtype.MakeLinearDNASequence("ColorDNA", seq)
+	for _, pc := range p.LivingColors{
+		if reflect.DeepEqual(pc,c) {
+			ok =  true
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go]: return true

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go [hunk 9]
```diff
@@ -1402,28 +1466,101 @@ func sqDiff(x, y uint32) uint32 {
...
+	}else{
+		return false
 	}
 
-	//populate livingColor
-	livingColor = &LivingColor{*color, DNASequence, component}
+}
 
-	return livingColor
+//Returns an array of unique state changes in a LivingGIF
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go]: godoc style on all exported functions.

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go [hunk 10]
```diff
@@ -1402,28 +1466,101 @@ func sqDiff(x, y uint32) uint32 {
...
+			stateList[stateNum] = append(stateList[stateNum], pix.Color.ID)
+			stateNum++
+		}
+
+		stateNum = 0
+	}
+
+	//remove duplicates
+	//concatenate state lists to use computationally efficient maps (this is necessary, using arrays takes way too long)
+	stateListNm := []string{}
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go]: var stateListNm []string

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go [hunk 11]
```diff
@@ -1402,28 +1466,101 @@ func sqDiff(x, y uint32) uint32 {
...
+        stateListNm[j] = stateNm
+        j++
+	}
+	stateListNm = stateListNm[:j]
+
+	//parsing resulting strings to arrays to make more sense in returned data
+	uniqueStates := [][]string{}
+
+	for _, state := range stateListNm{
+		var re = regexp.MustCompile(`\w+`)
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go]: lift out of loop

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate_test.go [hunk 12]
```diff
@@ -115,7 +120,7 @@ func TestMakeLivingImg(t *testing.T) {
 	}
 
 	//making the array to make palette. It's the same length than the array from the "ProteinPaintbox" library
-	for i := 1; i < 3; i++ {
+	for i := 1; i < 7; i++ {
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate_test.go]: len(livingColors)?

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate_test.go [hunk 13]
```diff
@@ -192,3 +197,150 @@ func TestMakeLivingGIF(t *testing.T) {
 
 	MakeLivingGIF(anthaImgs)
 }
+
+func TestAnthaPrintWorkflow(t *testing.T){
+
+	ctx := testinventory.NewContext(context.Background())
+
+	//downloading image for the test
+	imgFile, err := download.File("http://orig08.deviantart.net/a19f/f/2008/117/6/7/8_bit_mario_by_superjerk.jpg", "Downloaded file")
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate_test.go]: just store this image in a testdata directory

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate_test.go [hunk 14]
```diff
@@ -192,3 +197,150 @@ func TestMakeLivingGIF(t *testing.T) {
...
+}
+
+func TestOpenGIF(t *testing.T){
+
+	//downloading GIF for the test
+	//RAINBOWSPINNER
+	//https://media.giphy.com/media/XUHmgf1ij7dOU/source.gif
+	//BUTTERFLY
+	//https://www.google.co.uk/search?q=artsy+gif&source=lnms&tbm=isch&sa=X&ved=0ahUKEwi08uHk1szVAhUBAsAKHR6DAWUQ_AUICigB&biw=1301&bih=654#imgrc=iWSaQhyp0mvc5M:
+	GIFFile, err := download.File("https://media.giphy.com/media/XUHmgf1ij7dOU/source.gif", "Downloaded GIF")
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate_test.go]: ditto

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate_test.go [hunk 15]
```diff
@@ -192,3 +197,150 @@ func TestMakeLivingGIF(t *testing.T) {
...
+
+	//opening GIF
+	GIF, err := OpenGIF(GIFFile)
+	if err != nil {
+		t.Error(err)
+	}
+
+
+	//saving Image
+	fimg, err := os.Create("/home/cachemoi/gocode/src/github.com/cachemoi/GIF/doc/img/GIFImgTest.jpg")
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate_test.go]: remove hard-coded path

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate_test.go [hunk 16]
```diff
@@ -192,3 +197,150 @@ func TestMakeLivingGIF(t *testing.T) {
...
+	}
+	jpeg.Encode(fimg, imgs[0], &jpeg.Options{jpeg.DefaultQuality})
+
+}
+
+func TestGetState (t *testing.T){
+	ctx := testinventory.NewContext(context.Background())
+
+	lp := SelectLivingColorLibrary(ctx,"ProteinPaintBox")
+	lip,_  := inventory.NewPlate(ctx, "greiner384")
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate_test.go]: goft

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate_test.go [hunk 17]
```diff
@@ -192,3 +197,150 @@ func TestMakeLivingGIF(t *testing.T) {
...
+	fgif, err := os.Create("/home/cachemoi/gocode/src/github.com/cachemoi/GIF/doc/GIF/results.gif")
+	err = gif.EncodeAll(fgif,GIF)
+	if err != nil {
+		t.Error(err)
+	}
+}
+
+func TestParseGIF(t *testing.T) {
+
+	GIFFile, err := download.File("https://media.giphy.com/media/XUHmgf1ij7dOU/source.gif", "Downloaded GIF")
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate_test.go]: ditto and go style `gifFile`

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate_test.go [hunk 18]
```diff
@@ -192,3 +197,150 @@ func TestMakeLivingGIF(t *testing.T) {
...
+	lia = append(lia,*li1,*li2)
+	//------------------------------------------------
+	//Testing GIF functions
+	//------------------------------------------------
+
+	LivingGIF := MakeLivingGIF(lia)
+
+	states := LivingGIF.GetStates()
+
+	t.Log(states)
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate_test.go]: remove or replace with an actual assertion.

File: antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go [hunk 19]
```diff
@@ -1572,7 +1731,7 @@ func MakeLivingGIF(imgs []LivingImg) *LivingGIF {
 //Image manipulation
 //---------------------------------------------------
 
-//Minimalist resize function. Uses Lanczos resampling, which is the best but slowest method.
+// Minimalist resize function. Uses Lanczos resampling, which is the best but slowest method.
```
Comment (ddn0) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go]: godoc

Comment (cachemoi) [inline antha/AnthaStandardLibrary/Packages/image/pixeltoplate.go]: done, all my functions follow godoc convention now as far as I can tell

**Comment by chr1sgrant** [comment]
Fix failing tests

**Pull request merged**